### PR TITLE
Support aggregation of multiple NHD Regions

### PR DESCRIPTION
**Pull request by dtarb**
This supports delineation of watersheds that cross NHD Plus regions

### CODE REVIEW

**Comment by kdeloach** [comment]
This looks good, but we'll need to revert some of the Windows specific changes in order for this to work in our environment.

The best way to verify that these changes will work in production is by running the code inside the [docker-rwd](https://github.com/wikiwatershed/docker-rwd) container.

**Comment by dtarb** [comment]
I believe I have comments in the code at the only places I think you will need to revert (get_gage_watershed_command and generate_moveoutlettostream_command).

I do not know enough about Docker (and do not have Docker set up) that I could efficiently test in a Docker container, so will need to leave this to you.

**Comment by kdeloach** [comment]
Something else to consider is creating separate git commits for platform specific code changes. That would make it easier to revert those changes later. It isn't something we have to do for this pull request, but maybe future pull requests.

**Comment by dtarb** [comment]
The data processing for the Continental US (lower 48 states) is complete and ready to deploy.  It is 500 GB in 84259 subwatershed folders posted in 86 zip files (85 subwatersheds and one top level zip file) in https://drive.google.com/drive/folders/0B7V8il12WGQJM1JFWkN6bXpOX1k.  (Sorry I forgot to change the prefix MS before zipping, so do not be mislead.  The files are for the US, not only Mississippi region).

I addition to the data you should deploy the python code in this PR and update TauDEM MoveOutletsToStrm and gagewatershed functions from https://github.com/dtarb/TauDEM develop branch.

Gagewatershed has fixes in it that resolve incomplete handling of id field types that sometimes resulted in non-initialized and hence erratic id values in the outputs.  (This may not strictly be necessary for RWD as I do not think RWD id field types trigger this error, but it is better to have the bug fixed anyway).

MoveOutletsToStream has fixes for an infinite loop that occurred in certain cases where moving a snapped point crossed a parallel partition.  If a user clicks on the exact wrong point with the old code they could trigger this infinite loop (I did in my testing).

**Comment by kdeloach** [comment]
Do you mind creating a release for the version of TauDEM that you would like us to target? This will help avoid issues where changes to the develop branch slip into production before we've had a chance to test it.

**Comment by dtarb** [comment]
Sure, but it will on ly be next week.  I am at the AGU conference this week.

**Comment by dtarb** [comment]
I just created TauDEM 5.3.8 for you to use https://github.com/dtarb/TauDEM/releases/tag/v5.3.8.  This has been tested using the TauDEM test suite at https://github.com/dtarb/TauDEM-Test-Data which I updated to include tests for the new functions.

**Comment by kdeloach** [comment]
Thanks!

File: rwd_nhd/NHD_Rapid_Watershed_Delineation.py [hunk 1]
```diff
@@ -25,9 +26,11 @@ def Point_Watershed_Function(
 
     start_time = time.time()
 
-    dir_main = os.path.join(str(pre_process_dir), 'Main_Watershed')
+    dir_main = pre_process_dir
```
Comment (kdeloach) [inline rwd_nhd/NHD_Rapid_Watershed_Delineation.py]: For the sake of keeping these files grouped together, and to maintain parity with the DRB folder layout, I'd like to consider reverting this change.

Comment (dtarb) [inline rwd_nhd/NHD_Rapid_Watershed_Delineation.py]: I am OK with you organizing the folders as you deem best.

Comment (kdeloach) [inline rwd_nhd/NHD_Rapid_Watershed_Delineation.py]: Thanks.

**Comment by kdeloach** [comment]
@dtarb,

I noticed some differences in the `New_Point_Watershed.shp` output file compared to the previous version of NHD.

It looks like the field names of the output layer have changed (km suffixes added) and more importantly,  the values don't seem to be calculated correctly. As a result, we won't be able to display the area of the delineated watershed on the frontend.

The lat lng I used for this example is `39.972384 -75.201416`.

Before:
```
root@2da1e21f2ec8:/opt/rwd/src/api# ogrinfo -al -geom=NO /tmp/tmpQ_Jb3T/New_Point_Watershed.shp
INFO: Open of `/tmp/tmpQ_Jb3T/New_Point_Watershed.shp'
      using driver `ESRI Shapefile' successful.

Layer name: New_Point_Watershed
Metadata:
  DBF_DATE_LAST_UPDATE=2017-01-03
                -75.1790419727804,
                40.03518292440264
Geometry: Polygon
Feature Count: 1
Extent: (1632344.000000, 2074454.000000) - (1751026.000000, 2162956.000000)
Layer SRS WKT:
PROJCS["NAD_1983_Albers",
    GEOGCS["GCS_North_American_1983",
        DATUM["North_American_Datum_1983",
            SPHEROID["GRS_1980",6378137.0,298.257222101]],
        PRIMEM["Greenwich",0.0],
        UNIT["Degree",0.017453292519943295]],
    PROJECTION["Albers_Conic_Equal_Area"],
    PARAMETER["false_easting",0.0],
    PARAMETER["false_northing",0.0],
    PARAMETER["longitude_of_center",-96.0],
    PARAMETER["standard_parallel_1",29.5],
    PARAMETER["standard_parallel_2",45.5],
    PARAMETER["latitude_of_center",23.0],
    UNIT["Meter",1.0],
    VERTCS["Unknown VCS",
        VDATUM["Unknown"],
        PARAMETER["Vertical_Shift",0.0],
        PARAMETER["Direction",1.0],
        UNIT["User_Defined_Unit",0.01]]]
GRIDCODE: Integer (9.0)
Area: Real (24.15)
BasinLen: Real (24.15)
Strord: Real (24.15)
StrLen: Real (24.15)
DrnDen: Real (24.15)
AvgOLF: Real (24.15)
OGRFeature(New_Point_Watershed):0
  GRIDCODE (Integer) = 1
  Area (Real) = 4882.804199999999582
  BasinLen (Real) = 213.121437500000013
  Strord (Real) = 10.000000000000000
  StrLen (Real) = 195067.647999999986496
  DrnDen (Real) = 39.949922218875791
  AvgOLF (Real) = 0.012515668923224
```

After:
```
root@e07134c918b9:/opt/rwd/src/api# ogrinfo -al -geom=NO /tmp/tmp64V7bx/New_Point_Watershed.shp
INFO: Open of `/tmp/tmp64V7bx/New_Point_Watershed.shp'
      using driver `ESRI Shapefile' successful.

Layer name: New_Point_Watershed
Metadata:
  DBF_DATE_LAST_UPDATE=2017-01-03
Geometry: Polygon
Feature Count: 1
Extent: (1632344.000000, 2074454.000000) - (1751026.000000, 2162956.000000)
Layer SRS WKT:
PROJCS["Albers",
    GEOGCS["GCS_North_American_1983",
        DATUM["North_American_Datum_1983",
            SPHEROID["GRS_1980",6378137,298.257222101]],
        PRIMEM["Greenwich",0],
        UNIT["Degree",0.017453292519943295]],
    PROJECTION["Albers_Conic_Equal_Area"],
    PARAMETER["standard_parallel_1",29.5],
    PARAMETER["standard_parallel_2",45.5],
    PARAMETER["latitude_of_center",23],
    PARAMETER["longitude_of_center",-96],
    PARAMETER["false_easting",0],
    PARAMETER["false_northing",0],
    UNIT["Meter",1]]
GRIDCODE: Integer (9.0)
Area_km2: Real (24.15)
BasinL_km: Real (24.15)
Strord: Real (24.15)
StrLen_km: Real (24.15)
DrnDen_kmi: Real (24.15)
AvgOLF_km: Real (24.15)
OGRFeature(New_Point_Watershed):0
  GRIDCODE (Integer) = (null)
  Area_km2 (Real) = -999.000000000000000
  BasinL_km (Real) = -999.000000000000000
  Strord (Real) = -999.000000000000000
  StrLen_km (Real) = -999.000000000000000
  DrnDen_kmi (Real) = -999.00000000000000
```

**Comment by kdeloach** [comment]
Besides the area not being calculated, the polygons produced from the new NHD data seem to match the old version.

**Comment by dtarb** [comment]
Where the watersheds delineated are within a single grid processing unit of NHD plus the watershed properties are calculated.  Where the watersheds delineated span NHD grid processing units the watershed properties can not be calculated the way the preprocessing was done.  I set the code to return -999 in these cases as an indicator of no data.  I added the suffixes to be specific about the units of each quantity in cases where it was evaluated, and validated that the quantities were being correctly calculated for small watersheds that do not span processing units.  @aufdenkampe or someone responsible for front end appearance could comment on what is best to display on the front end.  It should also be possible to compute area from the delineated polygon, but not other watershed properties.

**Comment by aufdenkampe** [comment]
@dtarb, is there any way to create a work-around to calculate Area, etc. for these cases where the watershed spans multiple grid processing units? I think this would be an important feature. In the next month I'll be making a pitch to the World Wildlife Fund and some of there funders that we have the capabilities to deploy MMW for the entire globe (using 90 m SRTM, and with sufficient new funding). It would be great to have RWD working well for large watersheds in the USA.

**Comment by kdeloach** [comment]
This may be a non-issue after all. I assumed these properties were used on the frontend, but we actually use turf to calculate the polygon area in MMW. So there's nothing we need to do for this.

**Comment by kdeloach** [comment]
:+1: 

Looks good during my testing. Feel free to merge this in.

**Comment by aufdenkampe** [comment]
Awesome!

**Pull request merged**